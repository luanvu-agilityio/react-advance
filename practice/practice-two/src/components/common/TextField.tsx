import type { InputHTMLAttributes, ReactNode } from 'react'
import styled from 'styled-components'

export type IconPosition = 'left' | 'right'

export interface TextFieldProps
  extends Omit<InputHTMLAttributes<HTMLInputElement>, 'size'> {
  icon?: ReactNode
  iconPosition?: IconPosition
  variant?: 'default' | 'search'
}

const TextFieldContainer = styled.div<{
  $variant?: 'default' | 'search'
}>`
  position: relative;
  display: flex;
  align-items: center;
  width: 100%;
  height: 100%;
  background-color: ${(props) =>
    props.$variant === 'search' ? 'transparent' : 'var(--black-shade-4)'};
  padding: 10px 16px;
  border-radius: 12px;
`

const StyledInput = styled.input<{
  $hasLeftIcon?: boolean
  $hasRightIcon?: boolean
}>`
  width: 100%;
  height: 100%;
  padding-left: ${(props) => (props.$hasLeftIcon ? '2.5rem' : '0')};
  padding-right: ${(props) => (props.$hasRightIcon ? '2.5rem' : '0')};
  font-size: 14px;
  font-weight: var(--font-weight-regular);
  font-family: var(--font-family-secondary);
  background-color: transparent;
  border: none;
  color: var(--black-color-default);

  &::placeholder {
    color: var(--black-shade-2);
  }

  &:focus {
    outline: none;
  }
`

const IconContainer = styled.div<{
  $position: IconPosition
}>`
  position: absolute;
  ${(props) => (props.$position === 'left' ? 'left: 0.75rem;' : 'right: 0;')}
  display: flex;
  align-items: center;
  justify-content: center;
  height: 2.5rem;
  width: 2.5rem;
  color: #000;
`

const TextField = ({
  icon,
  iconPosition = 'right',
  variant = 'default',
  ...inputProps
}: TextFieldProps) => {
  return (
    <TextFieldContainer $variant={variant}>
      {icon && iconPosition === 'left' && (
        <IconContainer $position="left">{icon}</IconContainer>
      )}
      <StyledInput
        $hasLeftIcon={icon !== undefined && iconPosition === 'left'}
        $hasRightIcon={icon !== undefined && iconPosition === 'right'}
        {...inputProps}
      />
      {icon && iconPosition === 'right' && (
        <IconContainer $position="right">{icon}</IconContainer>
      )}
    </TextFieldContainer>
  )
}

export default TextField
