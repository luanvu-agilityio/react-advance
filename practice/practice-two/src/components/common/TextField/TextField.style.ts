import styled from 'styled-components'
import type { IconPosition } from './index'

export const TextFieldContainer = styled.div<{
  $variant?: 'default' | 'search'
}>`
  position: relative;
  display: flex;
  align-items: center;
  width: 100%;
  height: 100%;
  background-color: ${(props) =>
    props.$variant === 'search' ? 'transparent' : 'var(--black-shade-4)'};
  padding: 10px 16px;
  border-radius: 12px;
`

export const StyledInput = styled.input<{
  $hasLeftIcon?: boolean
  $hasRightIcon?: boolean
}>`
  width: 100%;
  height: 100%;
  padding-left: ${(props) => (props.$hasLeftIcon ? '2.5rem' : '0')};
  padding-right: ${(props) => (props.$hasRightIcon ? '2.5rem' : '0')};
  font-size: 14px;
  font-weight: var(--font-weight-regular);
  font-family: var(--font-family-secondary);
  background-color: transparent;
  border: none;
  color: var(--black-color-default);

  &::placeholder {
    color: var(--black-shade-2);
  }

  &:focus {
    outline: none;
  }
`

export const IconContainer = styled.div<{
  $position: IconPosition
}>`
  position: absolute;
  ${(props) => (props.$position === 'left' ? 'left: 0.75rem;' : 'right: 0;')}
  display: flex;
  align-items: center;
  justify-content: center;
  height: 2.5rem;
  width: 2.5rem;
  color: #000;
`
