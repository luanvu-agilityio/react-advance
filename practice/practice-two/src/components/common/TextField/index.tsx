'use client'

import type { InputHTMLAttributes, ReactNode } from 'react'
import {
  IconContainer,
  StyledInput,
  TextFieldContainer,
} from './TextField.style'

export type IconPosition = 'left' | 'right'

export interface TextFieldProps
  extends Omit<InputHTMLAttributes<HTMLInputElement>, 'size'> {
  icon?: ReactNode
  iconPosition?: IconPosition
  variant?: 'default' | 'search'
}

const TextField = ({
  icon,
  iconPosition = 'right',
  variant = 'default',
  ...inputProps
}: TextFieldProps) => {
  return (
    <TextFieldContainer $variant={variant}>
      {icon && iconPosition === 'left' && (
        <IconContainer $position="left">{icon}</IconContainer>
      )}
      <StyledInput
        $hasLeftIcon={icon !== undefined && iconPosition === 'left'}
        $hasRightIcon={icon !== undefined && iconPosition === 'right'}
        {...inputProps}
      />
      {icon && iconPosition === 'right' && (
        <IconContainer $position="right">{icon}</IconContainer>
      )}
    </TextFieldContainer>
  )
}

export default TextField
