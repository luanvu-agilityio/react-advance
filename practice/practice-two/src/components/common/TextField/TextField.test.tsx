import { render, screen, fireEvent } from '@testing-library/react'
import TextField from './index'
import { Search } from 'lucide-react'

describe('TextField Component', () => {
  const defaultProps = {
    placeholder: 'Enter text...',
    onChange: jest.fn(),
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  // SNAPSHOT TESTS FOR RENDERING
  describe('Rendering', () => {
    it('renders all variants correctly', () => {
      // Default rendering
      const { container: defaultContainer } = render(
        <TextField {...defaultProps} />
      )
      expect(defaultContainer).toMatchSnapshot('default')

      // With left icon
      const { container: leftIconContainer } = render(
        <TextField
          {...defaultProps}
          icon={<Search data-testid="search-icon" />}
          iconPosition="left"
        />
      )
      expect(leftIconContainer).toMatchSnapshot('left-icon')

      // With right icon
      const { container: rightIconContainer } = render(
        <TextField
          {...defaultProps}
          icon={<Search data-testid="search-icon" />}
          iconPosition="right"
        />
      )
      expect(rightIconContainer).toMatchSnapshot('right-icon')

      // Search variant
      const { container: searchContainer } = render(
        <TextField {...defaultProps} variant="search" />
      )
      expect(searchContainer).toMatchSnapshot('search-variant')

      // Disabled state
      const { container: disabledContainer } = render(
        <TextField {...defaultProps} disabled />
      )
      expect(disabledContainer).toMatchSnapshot('disabled')

      // With value
      const { container: withValueContainer } = render(
        <TextField {...defaultProps} value="Prefilled value" />
      )
      expect(withValueContainer).toMatchSnapshot('with-value')

      // Required field
      const { container: requiredContainer } = render(
        <TextField {...defaultProps} required />
      )
      expect(requiredContainer).toMatchSnapshot('required')
    })
  })

  // INTERACTIVE BEHAVIOR TESTS
  describe('Interactive behavior', () => {
    it('handles user input correctly', () => {
      render(<TextField {...defaultProps} />)
      const input = screen.getByPlaceholderText('Enter text...')

      fireEvent.change(input, { target: { value: 'test input' } })

      expect(defaultProps.onChange).toHaveBeenCalledTimes(1)
      expect(input).toHaveValue('test input')
    })

    it('handles focus and blur events', () => {
      const onFocus = jest.fn()
      const onBlur = jest.fn()

      render(<TextField {...defaultProps} onFocus={onFocus} onBlur={onBlur} />)
      const input = screen.getByPlaceholderText('Enter text...')

      fireEvent.focus(input)
      expect(onFocus).toHaveBeenCalledTimes(1)

      fireEvent.blur(input)
      expect(onBlur).toHaveBeenCalledTimes(1)
    })
  })
})
