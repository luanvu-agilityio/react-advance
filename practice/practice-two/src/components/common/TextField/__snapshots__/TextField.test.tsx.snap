// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Component Rendering renders all variants correctly: default 1`] = `
<div>
  <div
    class="sc-Qotzb fnjeBx"
  >
    <input
      class="sc-fYsHOw eRuMEp"
      placeholder="Enter text..."
    />
  </div>
</div>
`;

exports[`TextField Component Rendering renders all variants correctly: disabled 1`] = `
<div>
  <div
    class="sc-Qotzb fnjeBx"
  >
    <input
      class="sc-fYsHOw eRuMEp"
      disabled=""
      placeholder="Enter text..."
    />
  </div>
</div>
`;

exports[`TextField Component Rendering renders all variants correctly: left-icon 1`] = `
<div>
  <div
    class="sc-Qotzb fnjeBx"
  >
    <div
      class="sc-dKREkF cMJVcA"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-search"
        data-testid="search-icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21 21-4.34-4.34"
        />
        <circle
          cx="11"
          cy="11"
          r="8"
        />
      </svg>
    </div>
    <input
      class="sc-fYsHOw deiwLa"
      placeholder="Enter text..."
    />
  </div>
</div>
`;

exports[`TextField Component Rendering renders all variants correctly: required 1`] = `
<div>
  <div
    class="sc-Qotzb fnjeBx"
  >
    <input
      class="sc-fYsHOw eRuMEp"
      placeholder="Enter text..."
      required=""
    />
  </div>
</div>
`;

exports[`TextField Component Rendering renders all variants correctly: right-icon 1`] = `
<div>
  <div
    class="sc-Qotzb fnjeBx"
  >
    <input
      class="sc-fYsHOw hLAbFO"
      placeholder="Enter text..."
    />
    <div
      class="sc-dKREkF dfvOtJ"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-search"
        data-testid="search-icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21 21-4.34-4.34"
        />
        <circle
          cx="11"
          cy="11"
          r="8"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`TextField Component Rendering renders all variants correctly: search-variant 1`] = `
<div>
  <div
    class="sc-Qotzb fBJWKj"
  >
    <input
      class="sc-fYsHOw eRuMEp"
      placeholder="Enter text..."
    />
  </div>
</div>
`;

exports[`TextField Component Rendering renders all variants correctly: with-value 1`] = `
<div>
  <div
    class="sc-Qotzb fnjeBx"
  >
    <input
      class="sc-fYsHOw eRuMEp"
      placeholder="Enter text..."
      value="Prefilled value"
    />
  </div>
</div>
`;
